### ISSUE TITLE
About the loss. Sincerely, I would like to ask：

### ISSUE THREAD

**Issue by Struggle-Forever** [HobbitLong/SupContrast#116]
The purpose of contrast loss is to minimize the positive sample distance while maximizing the negative sample distance. However, I only find minimizing the distance of positive samples in this loss, and I don't see maximizing the distance of negative samples? Can you tell me which codes achieve the maximum negative sample distance?

**Comment by HobbitLong** [comment]
I think this [line](https://github.com/HobbitLong/SupContrast/blob/master/losses.py#L89) does it.

**Comment by Struggle-Forever** [comment]
> I think this [line](https://github.com/HobbitLong/SupContrast/blob/master/losses.py#L89) does it.

log_prob = logits - torch.log(exp_logits.sum(1, keepdim=True))  
The  `logits` denote all samples' distances and   `torch.log(exp_logits.sum(1, keepdim=True))`   denote the negative samples' distances  .

The `log_prob = logits - torch.log(exp_logits.sum(1, keepdim=True))`  denote the positive samples' distances   and the  `mean_log_prob_pos = (mask * log_prob).sum(1) / mask.sum(1)` denotes the average  positive loss . 

What confuses me is that this feels like it only minimizes the positive sample distance. The loss of maximizing negative samples is not in the final loss.

I feel like I'm not understanding something, can you help me?

**Comment by Struggle-Forever** [comment]
> I think this [line](https://github.com/HobbitLong/SupContrast/blob/master/losses.py#L89) does it.

I have understood it, thank you. A simplification operation is used here.

**Comment by Struggle-Forever** [comment]
> I think this [line](https://github.com/HobbitLong/SupContrast/blob/master/losses.py#L89) does it.
I still don't understand it. Please help me, thanks.

**Comment by Struggle-Forever** [comment]
> > I think this [line](https://github.com/HobbitLong/SupContrast/blob/master/losses.py#L89) does it.
> > I still don't understand it. Please help me, thanks.

he `log_prob ` denotes all samples' distances and the `mask * log_prob` can obtain the positive sample.   This means let all the sample distances do the numerator/denominator , then get the loss of positive samples by mask.   This time my understanding should be correct.

**Comment by RizhaoCai** [comment]
I also wonder that, at  [line](https://github.com/HobbitLong/SupContrast/blob/331aab5921c1def3395918c05b214320bef54815/losses.py#L89 )
![image](https://user-images.githubusercontent.com/37952703/208297254-768f4772-ea20-4f9e-a76d-8d7ddd056013.png)

Isn't that the distance between positive pairs are being the denominator because of `exp_logits.sum(1, keepdim=True)`?

**Comment by Haroon-Wahab** [comment]
This implementation of loss is incorrect and is not exact version of paper.
